File: nextcloud.py
"""Nextcloud News API implementation."""

import base64
import logging
import time

import dateutil.parser
import flask
import models
from utils import json_response


base_api = flask.Blueprint('nextcloud_base', __name__, url_prefix='/index.php/apps/news/api')
api = flask.Blueprint('nextcloud', __name__, url_prefix='/index.php/apps/news/api/v1-2')


@api.before_request
def authenticate():
    """Authenticate a Nextcloud News user."""
    auth = flask.request.authorization
    if auth:
        if models.authorized_user(auth.username, auth.password):
            return
    return flask.Response('Authentication Required', 401, headers={
            'WWW-Authenticate': 'Basic realm="Login Required"'})


@base_api.route('/')
def index():
    """Index page."""
    return json_response({'apiLevels': ['v1-2']})


@api.route('/user')
def user():
    """Get the status."""
    username = flask.request.authorization.username
    return json_response({
        'userId': username,
        'displayName': username,
        'lastLoginTimestamp': int(time.time()),
        'avatar': None
    })

@api.route('/version')
def version():
    """Get the version."""
    return json_response({'version': '6.0.5'})


@api.route('/status')
def status():
    """Get the status."""
    return json_response({'version': '6.0.5'})


@api.route('/feeds')
def feeds():
    """Get all feeds."""
    feeds = []
    for f in models.Feed.select():
        feeds.append({
            'id': f.id,
            'url': f.url,
            'title': f.name,
            'folderId': f.folder_id,
            'faviconLink': f.faviconLink,
            'link': f.link,
        })
    # TODO: Add newestItemId if there are new items
    return json_response({'feeds': feeds})


@api.route('/folders')
def folders():
    """Get all folders."""
    folders = []
    for f in models.Folder.select():
        folders.append({
            'id': f.id,
            'name': f.name
        })
    return json_response({'folders': folders})


def _item_to_json(i):
    pubDate = dateutil.parser.parse(str(i.pubDate))
    return {
        'id': i.id,
        'guid': i.guid,
        'guidHash': base64.encodebytes(i.guid.encode()).decode().strip(),
        'url': i.url,
        'title': i.title,
        'author': i.author,
        'pubDate': pubDate,
        'body': i.content,
        'feedId': i.feed_id,
        'unread': not i.read,
        'starred': i.starred,
        'lastModified': i.updated,
    }


@api.route('/items')
def items():
    """Get items."""
    batchSize = int(flask.request.args.get('batchSize', -1))
    offset = int(flask.request.args.get('offset', -1))
    getRead = flask.request.args.get('getRead', 'true') == 'true'
    oldestFirst = flask.request.args.get('oldestFirst', 'false') == 'true'
    id_ = int(flask.request.args['id'])
    type_ = int(flask.request.args['type'])
    items = []
    items_query = models.Item.select()
    if type_ == 0:
        items_query = items_query.where(models.Item.feed_id == id_)
    elif type_ == 1:
        feed_ids = [f.id for f in models.Feed.select(models.Feed.id).where(
            models.Feed.folder_id == id_)]
        items_query = items_query.where(models.Item.feed_id.in_(feed_ids))
    elif type_ == 2:
        items_query = items_query.where(models.Item.starred)
    elif type_ == 3:
        pass  # Get all items
    else:
        logging.exception(Exception('Unknown type: {0}'.format(type_)))
        flask.abort(500)
    items = [_item_to_json(i) for i in items_query]
    if offset > 0:
        items_query = items_query.where(models.Item.id <= offset)
    if not getRead:
        items_query = items_query.where(models.Item.read == False)
    if oldestFirst:
        items_query = items_query.order_by(models.Item.pubDate)
    else:
        items_query = items_query.order_by(models.Item.pubDate.desc())
    if batchSize > 0:
        items_query = items_query.limit(batchSize)
    return json_response({'items': items})


@api.route('/items/<int:item_id>/read', methods=['PUT'])
def item_read(item_id):
    """Mark an item as read."""
    i = models.Item.get_by_id(item_id)
    i.read = True
    i.save()
    return json_response({})


@api.route('/items/<int:item_id>/unread', methods=['PUT'])
def item_unread(item_id):
    """Mark an item as unread."""
    i = models.Item.get_by_id(item_id)
    i.read = False
    i.save()
    return json_response({})


@api.route('/feeds/<int:feed_id>/read', methods=['PUT'])
def feed_read(feed_id):
    """Mark items of a feed as read."""
    newestItemId = int(flask.request.args['newestItemId'])
    if feed_id == 0:
        models.Item.update(read=True).where(models.Item.starred).execute()
    else:
        models.Item.update(read=True).where(
            (models.Item.feed_id == feed_id) & (models.Item.id <= newestItemId)).execute()
    return json_response({})


@api.route('/items/updated')
def items_updated():
    """Get updated items."""
    lastModified = int(flask.request.args['lastModified'])
    type_ = int(flask.request.args['type'])
    id_ = int(flask.request.args['id'])
    items_query = None
    if type_ == 0:
        items_query = models.Item.select().where(
            (models.Item.feed_id == id_) & (models.Item.updated >= lastModified)
        )
    elif type_ == 1:
        feed_ids = [f.id for f in models.Feed.select(models.Feed.id).where(
            (models.Feed.folder_id == 1) & (models.Item.updated >= lastModified)
        )]
        items_query = models.Item.select().where(
            (models.Item.feed_id.in_(feed_ids)) & (models.Item.updated >= lastModified)
        )
    elif type_ == 2:
        items_query = models.Item.select().where(
            (models.Item.starred) & (models.Item.updated >= lastModified)
        )
    elif type_ == 3:
        items_query = models.Item.select().where(models.Item.updated >= lastModified)
    else:
        logging.exception(Exception('Unknown type: {0}'.format(type_)))
        flask.abort(500)
    items = [_item_to_json(i) for i in items_query]
    return json_response({'items': items})


@api.route('/items/read/multiple', methods=['PUT'])
def items_read_multiple():
    """Mark multiple items as read."""
    items = flask.request.get_json().get('items', [])
    models.Item.update(read=True).where(models.Item.id.in_(items)).execute()
    return json_response({})


@api.route('/items/unread/multiple', methods=['PUT'])
def items_unread_multiple():
    """Mark multiple items as unread."""
    items = flask.request.get_json().get('items', [])
    models.Item.update(read=False).where(models.Item.id.in_(items)).execute()
    return json_response({})


@api.route('/items/star/multiple', methods=['PUT'])
def items_star_multiple():
    """Mark multiple items as starred."""
    items = flask.request.get_json().get('items', [])
    for i in items:
        feed_id = int(i['feedId'])
        guidHash = i['guidHash']
        guid = base64.decodebytes(guidHash.encode()).decode()
        models.Item.update(starred=True).where(
            (models.Item.feed_id == feed_id) & (models.Item.guid == guid)
        ).execute()
    return json_response({})


@api.route('/items/unstar/multiple', methods=['PUT'])
def items_unstar_multiple():
    """Mark multiple items as unstarred."""
    items = flask.request.get_json().get('items', [])
    for i in items:
        feed_id = int(i['feedId'])
        guidHash = i['guidHash']
        guid = base64.decodebytes(guidHash.encode()).decode()
        models.Item.update(starred=False).where(
            (models.Item.feed_id == feed_id) & (models.Item.guid == guid)
        ).execute()
    return json_response({})

File: models/nextcloud.go
package models

import (
	"fmt"
	"time"
)

// JSONTime is a time.Time value formatted for JSON
type JSONTime time.Time

// MarshalJSON converts the JSONTime time.Time value to a Unix timestamp
func (t JSONTime) MarshalJSON() ([]byte, error) {
	return []byte(fmt.Sprintf("%d", time.Time(t).Unix())), nil
}

// NCFeed represents a RSS feed
type NCFeed struct {
	Name        string  `json:"name"`
	URL         string  `json:"url"`
	Title       *string `json:"title"`
	FolderID    int     `json:"folderId"`
	FavIconLink *string `json:"faviconLink"`
	Link        *string `json:"link"`
}

// NCFolder represents a named collection of feeds
type NCFolder struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

// NCItem represents a RSS feed Item
type NCItem struct {
	ID       int      `json:"id"`
	GUID     string   `json:"guid"`
	GUIDHash string   `json:"guidHash"`
	URL      string   `json:"url"`
	Title    string   `json:"title"`
	Author   string   `json:"author"`
	Content  string   `json:"body"`
	PubDate  JSONTime `json:"pubDate"`
	FeedID   int      `json:"feedId"`
	Unead    bool     `json:"unread"`
	Starred  bool     `json:"starred"`
	Updated  JSONTime `json:"lastModified"`
}

func feedsToNc(feeds []Feed) []NCFeed {
	var ncFeeds []NCFeed
	for _, f := range feeds {
		ncFeeds = append(ncFeeds, NCFeed{
			Name:        f.Name,
			URL:         f.URL,
			Title:       f.Title,
			FolderID:    f.FolderID,
			FavIconLink: f.FavIconLink,
			Link:        f.Link,
		})
	}
	return ncFeeds
}

func foldersToNc(folders []Folder) []NCFolder {
	var ncFolders []NCFolder
	for _, f := range folders {
		ncFolders = append(ncFolders, NCFolder{
			ID:   f.ID,
			Name: f.Name,
		})
	}
	return ncFolders
}

func itemsToNc(items []Item) []NCItem {
	var ncItems []NCItem
	for _, i := range items {
		ncItems = append(ncItems, NCItem{
			ID:       i.ID,
			GUID:     i.GUID,
			GUIDHash: i.GUIDHash,
			Title:    i.Title,
			Author:   i.Author,
			Content:  i.Content,
			PubDate:  JSONTime(i.PubDate),
			FeedID:   i.FeedID,
			Unead:    !i.Read,
			Starred:  i.Starred,
			Updated:  JSONTime(i.Updated),
		})
	}
	return ncItems
}

// NCGetAllFeeds gets all feeds for use with NextCloud
func NCGetAllFeeds() []NCFeed {
	return feedsToNc(getAllFeeds())
}

// NCGetAllFolders gets all folders for use with NextCloud
func NCGetAllFolders() []NCFolder {
	return foldersToNc(getAllFolders())
}

// NCQueryItems queries RSS items for use with NextCloud
func NCQueryItems() []NCItem {
	return itemsToNc(queryItems())
}
